src/librustc_trans/back/msvc/mod.rs

```diff
@@ -42,110 +44,133 @@ mod platform {
-            p.push(binsub);
-            p.push("link.exe");
-            if !p.is_file() { return None }
-            Some(Command::new(p))
-        }).or_else(|| {
-            get_vc_dir().and_then(|(ver, vcdir)| {
-                debug!("Found VC installation directory {:?}", vcdir);
-                let linker = vcdir.clone().join("bin").join(binsub).join("link.exe");
-                if !linker.is_file() { return None }
-                let mut cmd = Command::new(linker);
-                add_lib(&mut cmd, &vcdir.join("lib").join(vclibsub));
-                if ver == "14.0" {
-                    if let Some(dir) = get_ucrt_dir() {
-                        debug!("Found Universal CRT {:?}", dir);
-                        add_lib(&mut cmd, &dir.join("ucrt").join(libsub));
-                    }
-                    if let Some(dir) = get_sdk10_dir() {
-                        debug!("Found Win10 SDK {:?}", dir);
-                        add_lib(&mut cmd, &dir.join("um").join(libsub));
-                    } else if let Some(dir) = get_sdk81_dir() {
-                        debug!("Found Win8.1 SDK {:?}", dir);
-                        add_lib(&mut cmd, &dir.join("um").join(libsub));
-                    }
-                } else if ver == "12.0" {
-                    if let Some(dir) = get_sdk81_dir() {
+    // For MSVC 14 we need to find the Universal CRT as well as either the
+    // Windows 10 SDK or Windows 8.1 SDK.
+    fn find_msvc_14(arch: &str) -> Option<(Command, PathBuf)> {
+        get_vc_dir("14.0").and_then(|path| {
+            get_linker(&path, arch).into_iter().zip(lib_subdir(arch)).next()
+        }).and_then(|((mut cmd, host), sub)| {
+            get_ucrt_dir().map(|dir| {
+                debug!("Found Universal CRT {:?}", dir);
+                add_lib(&mut cmd, &dir.join("ucrt").join(sub));
+            }).and_then(|_| {
+                get_sdk10_dir().map(|dir| {
+                    debug!("Found Win10 SDK {:?}", dir);
+                    add_lib(&mut cmd, &dir.join("um").join(sub));
+                }).or_else(|| {
+                    get_sdk81_dir().map(|dir| {
                         debug!("Found Win8.1 SDK {:?}", dir);
-                        add_lib(&mut cmd, &dir.join("um").join(libsub));
-                    }
-                } else { // ver == "11.0"
-                    if let Some(dir) = get_sdk8_dir() {
-                        debug!("Found Win8 SDK {:?}", dir);
-                        add_lib(&mut cmd, &dir.join("um").join(libsub));
-                    }
-                }
-                Some(cmd)
-            })
-        }).unwrap_or_else(|| {
-            debug!("Failed to locate linker.");
-            Command::new("link.exe")
-        });
+                        add_lib(&mut cmd, &dir.join("um").join(sub));
+                    })
+                })
+            }).map(move |_| (cmd, host))
+        })
+    }
+
+    // For MSVC 12 we only need to find the Windows 8.1 SDK.
```

Comment:
Do you know if there's documentation for each of these MSVC versions about the filesystem layout? It'd be great to include some documentation here.